src/modelfree/scheduling.py

```diff
@@ -73,14 +106,84 @@ def get_value(self, frac_remaining):
         self.end_frac = end_frac
+        self.shaping_env = shaping_env
 
     def get_value(self, frac_remaining):
         if self.end_frac == 0:
             anneal_progress = 1.0
         else:
             anneal_progress = min(1.0, (1 - frac_remaining) / self.end_frac)
         return (1 - anneal_progress) * self.start_val + anneal_progress * self.end_val
+
+
+class ConditionalAnnealer(Annealer):
+    """Anneal the value depending on some condition."""
+    def __init__(self, start_val, end_val, decay_factor, thresh, window_size,
+                 min_wait, max_wait, operator, metric, shaping_env):
+        super().__init__(start_val, end_val)
+        self.decay_factor = decay_factor
+        self.thresh = thresh
+        self.window_size = window_size
+        self.min_wait = min_wait
+        self.max_wait = max_wait
+        self.operator = operator
+        self.metric = metric
+        self.shaping_env = shaping_env
+
+        self.current_param_val = start_val
+        self.last_num_episodes = 0
+
+    def set_shaping_env(self, shaping_env):
+        """Set the environment attribute since passing into constructor may not be possible"""
+        self.shaping_env = shaping_env
+
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        state['shaping_env'] = None
+        return state
+
+    @classmethod
+    def from_dict(cls, cond_config, shaping_env=None):
+        # provided to help keep track of arguments
+        constructor_config = {
+            'decay_factor': 0.98,
+            'thresh': 0,
+            'start_val': 1,
+            'end_val': 0,
+            'window_size': 1,
+            'min_wait': 1,
+            'max_wait': 10000,
+            'operator': operator.gt,
+            'metric': 'sparse',  # dense, length
+        }
+        if 'operator' in cond_config:
+            cond_config['operator'] = getattr(operator, cond_config['operator'])
+        trimmed_config = {}
+        for k, v in cond_config.items():
+            if k in constructor_config:
+                trimmed_config[k] = v
+        constructor_config.update(trimmed_config)
+        return cls(**constructor_config, shaping_env=shaping_env)
+
+    def get_value(self, frac_remaining):
+        if self.shaping_env is None:
+            raise ValueError("ConditionalAnnealer requires a RewardShapingVecWrapper to be set")
+
+        current_data = self.shaping_env.get_log_buffer_data()
+        if current_data is None:  # if we have zero episodes thus far
+            return self.current_param_val
+        # this is fine since we are doing deque.appendleft to put together these values
+        current_metric_data = itertools.islice(current_data[self.metric], self.window_size)
+        current_wait = current_data['num_episodes'] - self.last_num_episodes
+        if current_wait < self.min_wait:
+            return self.current_param_val
+
+        avg_metric_val = float(sum(current_metric_data)) / self.window_size
```

Comment:
Is `current_metric_data` always `self.window_size` long? (Can you get shorter slices e.g. at the end?)